BrailleNote: support for QT and Apex BT scroll wheel

josephsl opened the pull request:
This pull request addresses the following issues:
- #5992: Support for Apex BT's scroll wheel.
- #5993: Support for BraileNote QT (basic mode) when used as a braille display with NVDA.

All changes are documented in the user guide.

Suggested what's new:
Category: New features
Added support for BrailleNote QT and Apex BT's scroll wheel when BrailleNote is used as a braille display with NVDA.

Thanks.

josephsl committed:
BrailleNote: Support scroll wheel on Apex BT. re #5992

josephsl committed:
BrailleNote: Support BrailleNote QT basic mode. re #5993

josephsl committed:
BrailleNote: Various BT and QT commands now work correctly. re #5992.

Somehow, forgot to check that Rui used a tuple to specify key assignments. Now utilized properly to assign multiple gestures to a command. Also took this time to add codes for various QT keys so NVDA can properly recognize them.

josephsl committed:
Merge branch 'master' into i5992

jcsteh commented:
Thanks for your work. I have a few overall questions/concerns about this first before I do a formal review.

#### Keyboard Emulation

Is your intention to support keyboard emulation, but you weren't able to manage this yet? Or is there any reason you don't want to support this? Supporting that will require a change to the approach you're using. I think it might be worth considering that now, rather than having to do a major refactor later. Following are some thoughts on that.

Rather than emitting BrailleNote gestures and mapping them to key presses, I think it'd be simpler to just emit most keys as KeyboardInputGestures. The QT keys are actually Windows vk codes. So, for most key presses, you would do the following:

1. Map the QT modifiers to Windows vk codes; e.g. control to VK_CONTROL.
2. Instantiate a KeyboardInputGesture. Some notes on the arguments:
 * modifiers: The vk codes you mapped in step 1.
 * vkCode: The key code you got from the BrailleNote.
 * scanCode: None.
 * isExtended: Normally True, unless the vk code is 0xD (enter), in which case False.
3. Call inputCore.manager.emulateGesture with this gesture.

For most keys involving the QT function or read modifiers, you'd still want to produce BrailleNote specific gestures as you do now, as these are specific to the BrailleNote. To convert the vk code to a name, you can use KeyboardInputGesture.getVkName.

You might want to map control+read to the Windows key and then just treat it as a KeyboardInputGesture as above.

#### Handling Key Packets

At present, you read the extra bytes for QT_MOD_TAG in _readKeys. I suggest moving this to _readPacket, as this is where "reading" is done. Also, data is really just a second argument, so perhaps call it arg2.

josephsl commented:
Hi,

1. When to read packet: I'll do some experiments this weekend to see what's possible and moving the parsing routine to readPacket function.
2. QT emulation: what's making this confusing is the fact that Humanware has implemented PC keyboard emulation mode. Effectively, a QT user can switch between standard terminal mode (the one currently implemented) and PC keyboard mode where some combinations map directly to key presses.
3. QT terminal commands: My intention is to provide similar experience to that of BT users, but considering that people would like to type on QT keyboard and to serve as an example for other QWERTY keyboard notetakers, I'd be happy to let the branch support suggestions as outlined above (scan code/VK code entry).

Thanks.

jcsteh commented:
> When to read packet: I'll do some experiments this weekend to see what's possible and moving the parsing routine to readPacket function.

I'd have readPacket return three values, with the third being None for most commands except the QT modifier command.

> QT emulation: what's making this confusing is the fact that Humanware has implemented PC keyboard emulation mode. Effectively, a QT user can switch between standard terminal mode (the one currently implemented) and PC keyboard mode where some combinations map directly to key presses.

Yes, control mode and input mode according to the protocol documentation. I don't quite understand the point of input mode, though, because control mode also provides key presses. if anything, control mode is actually more "raw"; it gives you real modifiers. From what I've read, Window-Eyes allows QWERTY input in both, whereas JAWS only allows QWERTY input in input mode.

> 3. QT terminal commands: My intention is to provide similar experience to that of BT users, but considering that people would like to type on QT keyboard and to serve as an example for other QWERTY keyboard notetakers, I'd be happy to let the branch support suggestions as outlined above (scan code/VK code entry).

I wouldn't worry about serving as an example. If you don't feel this is appropriate, don't do it; it was just a suggestion because I thought people might like to be able to type. Different QWERTY notetakers do things differently anyway. For example, Papenmeier actually passes entirely raw key presses, right down to key ups and key downs for modifiers, whereas BrailleNote does special handling for modifiers.

josephsl committed:
BrailleNote QT: handle QT input directly from read packet function. re #5992

Comment from Jamie Teh (Nv Access): there's no need to call read packet again if we're dealing with QT, so handle this from read packet directly. Thus readPacket returns three things: command, arg (dot mask on BT/command for QT), arg2 (char from QT).

josephsl committed:
BrailleNote QT: QT key parser refactor.

Instead of relying on a dedicated QT key parser, handle parsing directly in readPacket. Also simplifies gesture ID assignment, as the packet will tell the gesture constructor as to what key it is.

jcsteh commented on source/brailleDisplayDrivers/brailleNote.py:
```diff
@@ -260,46 +318,54 @@ def display(self, cells):
+		# Make sure to display QT identifiers in mod+char format if this is such a case.
+		if self.qt:
+			self.id = qtData if qtMod == 0 else "+".join(("+".join(names), qtData))
+		else:
```
Gesture identifiers actually must be in Python set order; you can't reorder things. Otherwise, they might not match gesture maps. If you want to reorder them for display reasons, you will need to override `displayName` and `logIdentifier`. If you want to do that, you'll need a helper method which all of them call which returns the names in the correct order (list not set).

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -2093,6 +2100,39 @@ Please check your BrailleNote's documentation to find where these keys are locat
 | Toggle input help | space+dot2+dot3+dot6 (space+lower h) |
+
+Following are commands assigned to BrailleNote QT.
+%kc:beginInclude
```
1. A kc include block was already started earlier.
2. You probably want to include the sentence explaining the purpose of each table in the kc document anyway, so just leave the kc include block open from before.
3. Should the braille dot commands be split out of the main commands into a separate table, since I assume you can't execute those on a QT?

josephsl replied:
You can ask QT to emulate braille dots by pressing a key (it's a toggle). I can clarify this.

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -2093,6 +2100,39 @@ Please check your BrailleNote's documentation to find where these keys are locat
+| Toggle input help | read+1 |
+
+Following are commands assigned to the scroll wheel:
+%kc:beginInclude
```
See above re kc include block.

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -2093,6 +2100,39 @@ Please check your BrailleNote's documentation to find where these keys are locat
+%kc:beginInclude
+|| Name | Key |
+| NvDA menu | read+n |
+| Up arrow key | up arrow |
```
The convention in NVDA is normally to write keys in camel case; e.g. downArrow instead of down arrow.

jcsteh reviewed (changes requested):
Thanks!

Did you decide against doing QWERTY emulation and mapping vk codes using the existing `KeyboardInputGesture` stuff as I suggested? If so, that's fine, but be aware that emulation will require a significant refactor later.

josephsl committed:
BrailleNote QT: comment text change, correct ticket, simplified dictionary key lookup. re #5993

A tip from Jamie Teh (NV Access0: use dict.get to simplify the turnary operation. Also, BNQT ticket is 5993, not 5992, and correct comment text.

josephsl committed:
User guide: cosmetics. re #5992

Review from Jamie Teh (NV Access):
* QT can emulate a BrailleNote BT, hence clarify this.
* Removed redundant kcinclude statement.

josephsl committed:
BrailleNote QT: use set instead of worrying too much about formatting gesture identifier right. #5993

Comment from Jamie Teh (NV Access): use sets instead of turning that into a list just for formatting purposes (and turns out this is unnecessary).

leonardder commented:
This has now merge conflicts since this driver is hwIo based.

josephsl commented:
Hi, looks like I need to do a complete rebase. Thanks for telling me about this – will get to it next week (busy this weekend).

 

From: Leonard de Ruijter [mailto:notifications@github.com] 
Sent: Wednesday, November 8, 2017 12:21 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] BrailleNote: support for QT and Apex BT scroll wheel (#6316)

 

This has now merge conflicts since this driver is hwIo based.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6316#issuecomment-342946968> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkHHq3ZUlKbE30XkD0hRJrsseNtSwks5s0g0qgaJpZM4JwRDh> .

josephsl committed:
Merge master

josephsl committed:
Merge branch 'master' into i5992

# Conflicts:
#	source/appModules/win32calc.py
#	source/brailleDisplayDrivers/brailleNote.py

josephsl committed:
Readjust commands

feerrenrut commented:
@leonardder Would you mind reviewing this when you have a chance?

leonardder commented:
@josephsl: Could you please resolve another merge conflict?

LeonarddeR reviewed (changes requested):
Looks very good overall. Apart from the inline comments, could you please also revert the mode switch that occurred for appModules/win32calc.py

leonardder commented on source/brailleDisplayDrivers/brailleNote.py:
```diff
@@ -190,22 +235,33 @@ def _onReceive(self, command):
 			gesture = InputGesture(dots=arg, space=space)
+		elif command == QT_MOD_TAG:
+			# BrailleNote QT
+			gesture = InputGesture(qtMod=arg, qtData=arg2)
```
Have you actually considered making this a keyboardHandler.KeyboardInputGesture?

josephsl commented:
Hi, I will do so this weekend. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Wednesday, April 11, 2018 3:02 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] BrailleNote: support for QT and Apex BT scroll wheel (#6316)

 

@josephsl <https://github.com/josephsl> : Could you please resolve another merge conflict?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6316#issuecomment-380397878> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkJ4RSD6fNNJf2cqoZx0rW62IpL55ks5tndSCgaJpZM4JwRDh> .

josephsl commented:
Hi, not exactly, as I haven’t delved far into this. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Wednesday, April 11, 2018 3:33 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] BrailleNote: support for QT and Apex BT scroll wheel (#6316)

 

@leonardder commented on this pull request.

  _____  

In source/brailleDisplayDrivers/brailleNote.py <https://github.com/nvaccess/nvda/pull/6316#discussion_r180707076> :

>                              # Force dot8 here, although it should be already there
                               arg |= DOT_8
                       gesture = InputGesture(dots=arg, space=space)
+              elif command == QT_MOD_TAG:
+                      # BrailleNote QT
+                      gesture = InputGesture(qtMod=arg, qtData=arg2)

Have you actually considered making this a keyboardHandler.KeyboardInputGesture?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6316#pullrequestreview-111172061> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkI9NNmdmJCo7qmzve3DDwcPQO6mQks5tndvHgaJpZM4JwRDh> .

josephsl committed:
Merge branch 'master' into i5992

# Conflicts:
#	user_docs/en/userGuide.t2t

josephsl committed:
BrailleNote: copyright year update, QT key name changes. Re #5992.

Commented by Leonard de Ruijter: rename QT key names.
Update copyright years.

josephsl committed:
BrailleNote: rename scroll wheel constant strings. Re #5992.

leonardder commented:
@josephsl: Were you also intending to fix de mode change for win32calc.py?

josephsl commented:
Hi, ah yes, about to do that – chmod. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Friday, April 13, 2018 4:33 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] BrailleNote: support for QT and Apex BT scroll wheel (#6316)

 

@josephsl <https://github.com/josephsl> : Were you also intending to fix de mode change for win32calc.py?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6316#issuecomment-381108024> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkDPgLk9eO_CC0jIKtXcUt4lQAM37ks5toIzjgaJpZM4JwRDh> .

josephsl committed:
Restore executable permission to win32calc

derekriemer commented:
What's the status of this?

josephsl commented:
Hi, waiting for 2018.3 development to begin so we can come back to this. Thanks.

 

From: Derek Riemer <notifications@github.com> 
Sent: Sunday, June 10, 2018 1:59 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] BrailleNote: support for QT and Apex BT scroll wheel (#6316)

 

What's the status of this?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6316#issuecomment-396081502> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkKGNWwuNDc2OucAyztVjWKoI0YRoks5t7YidgaJpZM4JwRDh> .
